doc/man3/EZ_CreateWidget.3
'\"
'\" Copyright (c) 1997 Maorong Zou
'\" 
.TH EZ_CreateWidget 3 "" EZWGL "EZWGL Functions"
.BS
.SH NAME
EZ_CreateWidget, EZ_CreateWidgetXrm \- create an EZ widget
.SH SYNOPSIS
.nf
.B #include <EZ.h>
.sp
.BI "EZ_Widget *EZ_CreateWidget(int " type ", EZ_Widget *" parent ", " ... )
.sp
.BI "EZ_Widget *EZ_CreateWidgetXrm(int " type ", EZ_Widget *" parent,
.BI "            char *" name ", char *" className ", " ... )
.sp
.SH ARGUMENTS
\fItype\fR Specifies a symbolic widget type (see below). 
.sp
\fIparent\fR Specifies the parent widget or NULL.
.sp
\fIclassName\fR Specifies a class name.
.sp
\fIname\fR Specifies an instance name.
.sp
\fI ...\fR Specifies configuration options, a list of symbols 
each followed by one or more values. The last symbol must
be \fBNULL\fR, which is itself not followed by any values.
.sp
.SH DESCRIPTION
.PP
\fIEZ_CreateWidget\fR creates a widget and initializes the widget
using the specified configuration options. Then it reads the
resource database and configure the widget use resources
found in the database.
.PP
\fIEZ_CreateWidgetXrm\fR is equivalent to  \fIEZ_CreateWidget\fR 
with both \fBEZ_CLASS\fR and \fBEZ_NAME\fR attributes
specified.
.PP
Example 
.sp
.nf
   extern void exit();
   EZ_Widget *button;
                           /* type                   parent */
   button = EZ_CreateWidget(EZ_WIDGET_NORMAL_BUTTON, NULL,
                            /* option,       option value(s) */
                            EZ_LABEL_STRING, "Hello World",
                            EZ_UNDERLINE,    0,
                            EZ_CALLBACK,     exit, NULL,
                            0);
.sf
.sp
creates a push button widget labeled by "Hello World" with
the letter 'H' underlined.

.SH SYMBOLIC WIDGET TYPES 
.sp
.nf
EZ_WIDGET_FRAME,                  EZ_WIDGET_NORMAL_BUTTON, 
EZ_WIDGET_CHECK_BUTTON,            EZ_WIDGET_RADIO_BUTTON, 
EZ_WIDGET_MENU_BUTTON,           EZ_WIDGET_MENU_SEPARATOR, 
EZ_WIDGET_LABEL,                    EZ_WIDGET_RAW_XWINDOW, 
EZ_WIDGET_MENU_SUBMENU,      EZ_WIDGET_MENU_NORMAL_BUTTON, 
EZ_WIDGET_MENU_CHECK_BUTTON,  EZ_WIDGET_MENU_RADIO_BUTTON, 
EZ_WIDGET_VERTICAL_SLIDER,    EZ_WIDGET_HORIZONTAL_SLIDER, 
EZ_WIDGET_ENTRY,             EZ_WIDGET_VERTICAL_SCROLLBAR, 
EZ_WIDGET_HORIZONTAL_SCROLLBAR,        EZ_WIDGET_LIST_BOX,
EZ_WIDGET_TEXT,                       EZ_WIDGET_NOTE_BOOK, 
EZ_WIDGET_PANE_HANDLE,                 EZ_WIDGET_EXECUTOR, 
EZ_WIDGET_OPTIONAL_ENTRY,         EZ_WIDGET_FILE_SELECTOR, 
EZ_WIDGET_TREE,                       EZ_WIDGET_LIST_TREE, 
EZ_WIDGET_FANCY_LIST_BOX,             EZ_WIDGET_WORK_AREA, 
EZ_WIDGET_FREE_LABEL,                 EZ_WIDGET_3D_CANVAS,
EZ_WIDGET_NB_PAGE,                     EZ_WIDGET_NW_LABEL,
EZ_WIDGET_ICON,                             EZ_WIDGET_LCD,
EZ_WIDGET_LED,                     EZ_WIDGET_STATUS_METER,
EZ_WIDGET_MENU_BAR                         EZ_WIDGET_TERM,
EZ_WIDGET_HORIZONTAL_RULER,      EZ_WIDGET_VERTICAL_RULER,
EZ_WIDGET_SCROLL_BUTTON,            EZ_WIDGET_SPIN_BUTTON,
EZ_WIDGET_LOCATOR,                         EZ_WIDGET_DIAL,
EZ_WIDGET_GRADIENT_BAR,               EZ_WIDGET_HISTOGRAM,
EZ_WIDGET_SPREAD_SHEET,             EZ_WIDGET_THUMB_WHEEL,
.sf
.PP


.SH CONFIGURATION SYMBOLS AND THEIR VALUES

See manual for a complete list of resources/configuration
options.


.SH "SEE ALSO"
EZ_DisplayWidget(3), EZ_ConfigureWidget(3), EZ_DestroyWidget(3)




ChangeLog
Fri Dec 10 23:44:16 1999  m z  <zou@hitachi1.ma.utexas.edu>
        
        * Added EZ_SSheetSortRegionR, EZ_SSheetSortColumns/Rows
        
1999-12-04  Maorong Zou  <mzou@linux115.ma.utexas.edu>

        * Released version 1.50 
        * added a CVS repository entry 
        *  :pserver:anonymous@cvs.math.utexas.edu:/home/cvs/EZWGL
        * and a read-only cvs web interface
        *  http://cvs.math.utexas.edu/cgi-bin/cvswe.cgi?cvsroot=EZWGL

1999-11-18  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* Added THUMB_WHEEN and SPREAD_SHEET widget
	* a few modifications in in FigItem ...

1999-09-22  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* EZ_RESOURCE_STRING resource
	* EZ_LedLoadPixelMatrix() EZ_LedLoadOnOffMatrix()
	* EZ_AllowDualLabel()
	* EZ_FlashWidgetBorder2()
	* NewFile  EZ_ResourceEdit.c
	* Incompatible change:  EZ_LED_PIXEL_SIZE now
	*        takes a pair of ints (xsize, ysize)

	
Sun Aug 22 01:26:59 1999  m z  <mzou@crab.ma.utexas.edu>

        * Separator Widget

Tue Aug 17 02:10:03 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added   EZ_SlideOpenWidget,  EZ_SlideCloseWidget
        *         EZ_SlideOpenWidgetRelativeTo
        * new widget EZ_WIDGET_HISTOGRAM

Tue Aug 10 12:33:26 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added the pare EZ_FreezeDisplay() and EZ_ThawDisplay()
        * New File EZ_ResourceEdit.c (resource editor functions)
        * Added EZ_RegisterApplicationResources(n,spec,config)
        * Modified all resource handling functions
        * changed EZ_Initialize and EZ_InitializeXrm from void to int
        *   (return -1 on failure, 0 on success)
        
Mon Jul 12 02:13:37 1999  m z  <mzou@crab.ma.utexas.edu>

        * New widget GradientBar
        * Fix a bug in EZ_PixelValue2RGB
        
Tue Jul  6 02:17:29 1999  m z  <mzou@crab.ma.utexas.edu>

        * Fix buffer overflow bug in EZ_Xpm.c
        * added line/word wrap for textWidget, IN BROWSING FILE MODE ONLY!!
        * (need to rewrite textWidget, the current line based
        *  implementation is a design flaw)

Thu Jun 17 01:49:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * New resources: SHAPE_PIXMAP, SHAPE_FILE
        * EZ_Bitmap *EZ_CreateShapeMaskFromFigure(fitem) EZ_Item *fitem;
        * Pixmap EZ_MakePixmapFromFigItem(item, width, height)
             EZ_Item *item; int *width, *height;
        
Fri May 14 23:30:19 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added a new tree type: branched-list-tree

Thu May 13 01:50:55 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Widget:  ScrollButton, SpinButton, Locator

Sun May  9 02:03:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added UDF widget resource handling.
        * EZ_RegisterResourcesAll(type, nresrc, resrc, configure);
        * EZ_RegisterResourcesWidget(widget, nresrc, resrc, configure);        

Tue May  4 23:36:31 1999  m z  <mzou@crab.ma.utexas.edu>

        *  Widget (horizontal/vertical) ruler
        *  
        *  Configuration options:
        *  EZ_HIGHLIGHT_BACKGROUND, EZ_HIGHLIGHT_FOREGROUND,
        *  EZ_OVERRIDE_REDIRECT

Tue May  4 01:22:57 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Function 
        * void  EZ_ConstrainFreeLabel(EZ_Widget *widget, int flag);
        * void  EZ_MoveWidgetWindow(EZ_Widget *widget, int x, int y);
        * void  EZ_ResizeWidgetWindow(EZ_Widget *widget, int w, int h);
        * void  EZ_MoveResizeWidgetWindow(EZ_Widget *widget, int x, int y, int w, int h);        
        * Added optimization for manipulating images.
        *
        
Sat Apr 17 01:25:25 1999  m z  <mzou@crab.ma.utexas.edu>

        * Autoconf (finally)

1999-04-16  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * added a vt100 emulator widget (EZ_WIDGET_TERM), derived from xiterm.
        * added support for image format: png and tiff
        
Tue Dec  8 11:39:04 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        *  new function:     EZ_ScaleRGBT
        *                    EZ_XPixmap2RGB
        *                    EZ_SelectInput
        *                    EZ_CreateBackgroundPixmapsFromRawRGB

Sat Dec  5 03:06:58 1998  m z  <mzou@bob.ma.utexas.edu>

        * EZ_LABEL_SHADOW  "labelShadow" config. option

Thu Dec  3 14:54:21 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * new function:   EZ_SetDirTreeSorter
        *                 EZ_MarkWidgetDirty
        
Mon Nov 30 12:25:50 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * EZ_WM_IGNORE  "wmIgnore"    configuration option
        *
        * new function    EZ_GetIconTextGeometry
        *                 EZ_SetGlobalEventFilter(filter, data)
        
Thu Nov 19 16:35:47 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	*  EZ_TILE_ORIGIN  tileOrigin: mode x y
	*  resources.  mode=0,1,2    (0 for off)

	*  It is now possible to make all EZ widgets
	*  pseudo transparent under Enlightment WM ...

	*  A few image readers have been modified to
	*  return RGB images.

	*  new functions EZ_ScaleRGB, EZ_ReadImageFile,
	*                EZ_ScaleLabelPixmap,
	*                EZ_SwapLabelPixmaps, 
	*                EZ_CopyLabelPixmaps,
	*                EZ_TraverseAllWidgets,

	*  In EZ_DisplayWidget, the order for displaying widget tree
	*  is reversed. Now widgets refresh a lot faster visually.
	
	*  Bug Fix:  whe a radio button is destroyed, the widget is
	*            not removed from its button group immediately.



lib/EZ_CreateWidget.c
/*********************************************************************
 *
 *         EZWGL, the EZ Widget and Graphics Library
 *
 * *             Copyright (C) 1996, 1997, 1998  Maorong Zou
 *  
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 **********************************************************************/
/*
 *  June 1996.  Beta Release.
 *  Sept 1996.  Release Version 1.0
 *  Dec 1996.  Release Version 1.1 Beta
 *  April 1997.  Release Version 1.2
 *  November 1997.  Release Version 1.3
 *  November 1998.  Release Version 1.4
 *  December 1999.  Release Version 1.50
 */
/*****************************************************************
 ***                                                           ***
 ***              Creating Widgets                             ***
 ***                                                           ***
 *****************************************************************/
#define _EZ_CREATE_WIDGET_C_
#include "EZ_Widget.h"
/*****************************************************************/
static EZ_Widget *newWidget MY_ANSIARGS((EZ_Widget *parent, int type,
                                         char *iname, char *cname));

extern int    EZ_CounterUT;
/*****************************************************************/
static void fixNBPage(widget) EZ_Widget *widget;
{
  char tmp[256], *tname;
  EZ_Widget *tab = EZ_NBPageCompanion(widget);
  tname = XrmQuarkToString(EZ_WidgetIName(widget));
  strcpy(tmp, tname); strcat(tmp, "Tab");
  EZ_WidgetIName(tab) = EZ_StringToQuark(tmp);
  tname = XrmQuarkToString(EZ_WidgetCName(widget));
  strcpy(tmp, tname); strcat(tmp, "Tab");
  EZ_WidgetCName(tab) = EZ_StringToQuark(tmp);
}
/*****************************************************************/

EZ_Widget *EZ_CreateWidgetXrm MY_VARARGS(int, the_args)
     MY_VARARGS_DECL
{
  EZ_Widget *parent, *widget;
  char      *iname,  *cname;
  XrmQuark  qiname,  qcname; 
  int       type;
  va_list   ap;  
  
  /* if(EZ_Display == NULL) return(NULL);*/
  type = (int)( MY_VARARGS_START(int, the_args, ap));
  parent = (EZ_Widget *)va_arg(ap, void *);
  iname = (char *)va_arg(ap, char *);
  cname = (char *)va_arg(ap, char *);
  if(iname != NULL) qiname = EZ_StringToQuark(iname);
  else qiname = NULLQUARK;
  if(cname != NULL) qcname = EZ_StringToQuark(cname);
  else qcname = NULLQUARK;
  
  if( (widget = newWidget(parent, type, iname, cname)) == NULL)
    {  va_end(ap); return(NULL);}

  /* set resources from app-defaults */
  if(qiname != NULLQUARK) EZ_WidgetIName(widget) = qiname;
  if(qcname != NULLQUARK) EZ_WidgetCName(widget) = qcname;
  
  /* set widget attributes. These are considered as fallback resources */
  {
    int     i1,i2,i3,i4;
    void    *p1, *p2;
    unsigned long ul1, ul2;
    float   f1,f2;
    int     option, done = 0, size_changed = 0;
#include "EZ_WidgetAttr.h"    
  }

  if(type == EZ_WIDGET_NB_PAGE) fixNBPage(widget) ;

  EZ_GetAndUseWidgetResources(widget);
  return(widget);
}
/**********************************************************************/

EZ_Widget *EZ_CreateWidget MY_VARARGS(int, the_args)
     MY_VARARGS_DECL
{
  EZ_Widget *parent, *widget;
  int type;
  va_list   ap;  

  /* if(EZ_Display == NULL) return(NULL);*/
  
  type = (int)( MY_VARARGS_START(int, the_args, ap));
  parent = (EZ_Widget *)va_arg(ap, void *);

  if( (widget = newWidget(parent, type, NULL, NULL)) == NULL)
    {  va_end(ap); return(NULL);}
  /* set widget attributes. These are considered as fallback resources */
  {
    int     i1,i2,i3,i4;
    void    *p1, *p2;
    unsigned long ul1, ul2;
    float   f1,f2;
    int     option, done = 0, size_changed = 0;
#include "EZ_WidgetAttr.h"   
  }

  if(type == EZ_WIDGET_NB_PAGE) fixNBPage(widget) ;

  /* set resources from app-defaults */
  EZ_GetAndUseWidgetResources(widget);
  return(widget);
}
/************************************************************************************/  

static EZ_Widget *newWidget(parent, type, iname, cname)
     EZ_Widget *parent; int type;
     char *iname, *cname;
{
  EZ_Widget *widget = NULL;


  switch(type)
    {
    case EZ_WIDGET_FRAME:
    case EZ_WIDGET_ROW_COLUMN:  /* grid bag is obsolete */
      widget = EZ_CreateFrame(parent, NULL);
      break;
    case EZ_WIDGET_LABEL:
      widget = EZ_CreateLabel(parent,NULL);
      break;
    case EZ_WIDGET_ARROW_BUTTON:
      widget = EZ_CreateArrowButton(parent, 13);
      break;
    case EZ_WIDGET_NORMAL_BUTTON:
      widget = EZ_CreateButton(parent, NULL, -1);
      break;
    case EZ_WIDGET_CHECK_BUTTON:
      widget = EZ_CreateCheckButton(parent, NULL, -1, 1, 0, 0);
      break;
    case EZ_WIDGET_RADIO_BUTTON:
      widget = EZ_CreateRadioButton(parent, NULL, -1, 
				    (0x12345 + EZ_CounterUT++), (0x12345 + EZ_CounterUT++));
      break;
    case EZ_WIDGET_MENU_BUTTON:
      widget = EZ_CreateMenuButton(parent, NULL, -1);
      break;
    case EZ_WIDGET_MENU_SEPARATOR:
      widget = EZ_CreateMenuSeparator(parent);
      break;
    case EZ_WIDGET_MENU_SUBMENU:
      widget = EZ_CreateMenuSubMenu(parent, NULL, -1);
      break;
    case EZ_WIDGET_MENU_NORMAL_BUTTON:
      widget = EZ_CreateMenuNormalButton(parent, NULL, -1, 0);
      break;
    case EZ_WIDGET_MENU_CHECK_BUTTON:
      widget = EZ_CreateMenuCheckButton(parent, NULL, -1, 1, 0, 0, 0);
      break;
    case EZ_WIDGET_MENU_RADIO_BUTTON:
      widget = EZ_CreateMenuRadioButton(parent, NULL, -1, 
					(0x12345 + EZ_CounterUT++), (0x12345 + EZ_CounterUT++), 0);
      break;
    case EZ_WIDGET_VERTICAL_SLIDER:
      widget = EZ_CreateSlider(parent, NULL, 0.0, 1.0, 0.0, EZ_WIDGET_VERTICAL_SLIDER);
      break;
    case EZ_WIDGET_HORIZONTAL_SLIDER:
      widget = EZ_CreateSlider(parent, NULL, 0.0, 1.0, 0.0, EZ_WIDGET_HORIZONTAL_SLIDER);
      break;
    case EZ_WIDGET_ENTRY:
      widget = EZ_CreateEntry(parent, NULL);
      break;
    case EZ_WIDGET_3D_CANVAS:
      widget = EZ_Create3DCanvas(parent);
      break;
    case EZ_WIDGET_VERTICAL_SCROLLBAR:
      widget = EZ_CreateScrollbar(parent, EZ_WIDGET_VERTICAL_SCROLLBAR);
      break;
    case EZ_WIDGET_HORIZONTAL_SCROLLBAR:
      widget = EZ_CreateScrollbar(parent, EZ_WIDGET_HORIZONTAL_SCROLLBAR);
      break;
    case EZ_WIDGET_ILIST_BOX:
      widget = EZ_CreateIListBox(parent);
      break;
    case EZ_WIDGET_LIST_BOX:
      widget = EZ_CreateListBox(parent, 1, 1);
      break;
    case EZ_WIDGET_ITEXT:
      widget = EZ_CreateIText(parent,1);
      break;
    case EZ_WIDGET_TEXT:
      widget = EZ_CreateTextWidget(parent, 1, 1, 1);
      break;
    case EZ_WIDGET_NOTE_BOOK:
      widget = EZ_CreateNoteBook(parent);
      break;
    case EZ_WIDGET_NB_PAGE:
      widget = EZ_CreateNoteBookPage(parent, NULL, -1);
      break;
    case EZ_WIDGET_NW_LABEL:
      widget = EZ_CreateNWLabel(parent, NULL);
      break;
    case EZ_WIDGET_PANE_HANDLE:
      widget = EZ_CreatePaneHandle(parent);
      break;
    case EZ_WIDGET_EXECUTOR:
      widget = EZ_CreateExecutor(parent);
      break;
    case EZ_WIDGET_OPTIONAL_ENTRY:
      widget = EZ_CreateOptionalEntry(parent, NULL,1,1);
      break;
    case EZ_WIDGET_FILE_SELECTOR:
      widget = EZ_CreateFileSelector(parent, NULL);
      break;
    case EZ_WIDGET_ITREE:
      widget = EZ_CreateITree(parent);
      break;
    case EZ_WIDGET_TREE:
      widget = EZ_CreateListTree(parent, 1,1);
      break;
    case EZ_WIDGET_IFANCY_LIST_BOX:
      widget = EZ_CreateIFListBox(parent, 1, 0);
      break;
    case EZ_WIDGET_FANCY_LIST_BOX:
      widget = EZ_CreateFancyListBox(parent,1,1,1,0);
      break;
    case EZ_WIDGET_IWORK_AREA:
      widget = EZ_CreateIWorkArea(parent);
      break;
    case EZ_WIDGET_WORK_AREA:
      widget = EZ_CreateWorkArea(parent,1,1);
      break;
    case EZ_WIDGET_POPUP_MENU:
      widget = EZ_CreatePopupMenu(NULL);
      break;
    case EZ_WIDGET_ITOPLEVEL:
      widget = EZ_CreateIToplevelFrame(parent, NULL);
      break;
    case  EZ_WIDGET_EMBEDER:
      widget = EZ_CreateEmbeder(parent);
      break;
    case EZ_WIDGET_RAW_XWINDOW:
      widget = EZ_CreateRawXWindow(parent);
      break;
    case EZ_WIDGET_MENU_TEAR_OFF_BAR:
      widget = EZ_CreateMenuTearOffItem(parent);
      break;   
    case EZ_WIDGET_LCD:
      widget = EZ_CreateLcd(parent, 6);
      break;
    case EZ_WIDGET_LED:
      widget = EZ_CreateLed(parent);
      break;
    case EZ_WIDGET_STATUS_METER:
      widget = EZ_CreateStatusMeter(parent);
      break;      
    case EZ_WIDGET_MENU_BAR:
      widget = EZ_CreateMBar(parent);
      break;
    case EZ_WIDGET_ITERM:
      widget = EZ_CreateITerm(parent);
      break;
    case EZ_WIDGET_TERM:
      widget = EZ_CreateTerm(parent,1);
      break;
    case EZ_WIDGET_HORIZONTAL_RULER:   
      widget=EZ_CreateRuler(parent,EZ_WIDGET_HORIZONTAL_RULER);
      break;
    case EZ_WIDGET_VERTICAL_RULER:   
      widget=EZ_CreateRuler(parent,EZ_WIDGET_VERTICAL_RULER);
      break;
    case EZ_WIDGET_SCROLL_BUTTON:
      widget=EZ_CreateScrollButton(parent, EZ_UP_DOWN);
      break;
    case EZ_WIDGET_SPIN_BUTTON:
      widget=EZ_CreateSpinButton(parent);
      break;
    case EZ_WIDGET_LOCATOR:
      widget=EZ_CreateLocator(parent);
      break;
    case EZ_WIDGET_DIAL:
      widget=EZ_CreateDial(parent);      
      break;
    case EZ_WIDGET_GRADIENT_BAR:
      widget=EZ_CreateGradientBar(parent);      
      break;      
    case EZ_WIDGET_HISTOGRAM:
      widget=EZ_CreateHistogram(parent);      
      break;
    case EZ_WIDGET_SEPARATOR:
      widget=EZ_CreateSeparator(parent);      
      break;
    case EZ_WIDGET_SPREAD_SHEET: 
      widget=EZ_CreateSpreadSheet(parent);           
      break;
    case EZ_WIDGET_THUMB_WHEEL:
      widget=EZ_CreateThumbWheel(parent);                 
      break;
    default:
      widget = EZ_CreateUDFWidget(type, parent, iname, cname);
      break;
    }
  return(widget);
}
/**********************************************************************************************/
#undef _EZ_CREATE_WIDGET_C_



